presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java

```diff
@@ -13,35 +13,42 @@
  */
 package com.facebook.presto.spi.function;
 
-import com.facebook.presto.spi.QueryId;
 import com.facebook.presto.spi.api.Experimental;
 import com.facebook.presto.spi.relation.FullyQualifiedName;
 
 import java.util.Collection;
+import java.util.Optional;
 
 @Experimental
 public interface FunctionNamespaceManager<F extends SqlFunction>
 {
-    void createFunction(F function, boolean replace);
+    String getName();
 
-    Collection<F> listFunctions();
+    /**
+     * Start a transaction.
+     */
+    FunctionNamespaceTransactionHandle beginTransaction();
 
     /**
-     * Ideally function namespaces should support transactions like connectors do, and getCandidates should be transaction-aware.
-     * queryId serves as a transaction ID before proper support for transaction is introduced.
-     * TODO Support transaction in function namespaces
+     * Commit the transaction. Will be called at most once and will not be called if
+     * {@link #rollback(FunctionNamespaceTransactionHandle)} is called.
      */
-    Collection<F> getFunctions(QueryId queryId, FullyQualifiedName functionName);
+    void commit(FunctionNamespaceTransactionHandle transactionHandle);
 
     /**
-     * If a SqlFunction for a given signature is returned from {@link #getFunctions(QueryId, FullyQualifiedName)}
-     * for a given queryId, getFunctionHandle with the same queryId should return a valid FunctionHandle, even if the function
-     * is deleted. Multiple calls of this function with the same parameters should return the same FunctionHandle.
-     * queryId serves as a transaction ID before proper support for transaction is introduced.
-     * TODO Support transaction in function namespaces
-     * @return FunctionHandle or null if the namespace manager does not manage any function with the given signature.
+     * Rollback the transaction. Will be called at most once and will not be called if
+     * {@link #commit(FunctionNamespaceTransactionHandle)} is called.
+     * Note: calls to this method may race with calls to the ConnectorMetadata.
      */
-    FunctionHandle getFunctionHandle(QueryId queryId, Signature signature);
+    void rollback(FunctionNamespaceTransactionHandle transactionHandle);
+
+    void createFunction(F function, boolean replace);
```

Comment:
`createFunction` should probably also take a `FunctionNamespaceTransactionHandle` and support transaction. If a transaction is rolled back, the functions created / updated within that transaction should be discarded. You can do that in a follow up commit / PR though. Just add a TODO is fine for now.

Reply:
Left a TODO.